File: macos/Sources/TaigiInputMethodCore/Settings/EngineSettings.swift
// The settings one composing operation reads. The provider that serves them
// lives in EngineSettingsProvider.swift.

import Foundation

/// The romanization the user types. macOS ships TL and POJ only — TPS is
/// deliberately out of scope for this platform (`docs/architecture/macos-roadmap.md`
/// § Goal), which is why this enum has no `tps` case to fall through.
enum InputMode: String, CaseIterable, Sendable {
    case tl
    case poj
}

/// How the candidate window renders the `(Hanji, romanization)` pair: both scripts
/// side by side (the swap setting decides which leads), each script as its own
/// adjacent cell (Hanji with Romanization, `PresentedCandidate`), or the romanization alone.
///
/// Raw values are the storage contract every platform shares
/// (`docs/reports/2026-08-30-hanlo-together-mode-research.md` §12) — the same
/// convention `isTranslateSwapped` / `outputBothScripts` follow, so a future
/// settings transfer carries one vocabulary.
/// CROSS-PLATFORM INVARIANT — mirrors ios/Sources/TaigiKeyboard/Settings/SettingsModels.swift
/// `CandidateDisplayMode` and the Android `CandidateDisplayMode.storageValue`.
/// Drift changes which mode a transferred setting resolves to.
enum CandidateDisplayMode: String, CaseIterable, Sendable {
    case sideBySide
    case combined
    case romanOnly

    /// Whether the cell shows any Hanji — `false` only under `.romanOnly`.
    var showsHanji: Bool {
        self != .romanOnly
    }

    /// The mode after this one, in the order the Appearance picker lists them, and
    /// round again from the end — what the cycle shortcut steps through, so
    /// the key and the picker agree on what "next" is.
    var next: CandidateDisplayMode {
        let all = Self.allCases
        let index = all.firstIndex(of: self) ?? all.startIndex
        return all[(index + 1) % all.count]
    }

    /// The picker row's label for this mode — also what the cycle shortcut
    /// flashes, so the HUD names the mode in the words the pane uses.
    var displayNameKey: StringKey {
        switch self {
        case .sideBySide: .settingsCandidateDisplayModeSideBySide
        case .combined: .settingsCandidateDisplayModeCombined
        case .romanOnly: .settingsCandidateDisplayModeRomanOnly
        }
    }

    /// Whether the swap shortcut writes the stored swap — exactly where Hanji
    /// is on screen. Under `.combined` the cells are split per script, so the
    /// shortcut only picks the punctuation width (USER 2026-09-13: "Hanji with Romanization needs an
    /// isTranslateSwapped button"); `.romanOnly` leaves it inert and the stored
    /// swap waits for the way back.
    var allowsSwapToggle: Bool {
        showsHanji
    }

    /// Effective swap for a stored flag. `.combined` leads with — and commits —
    /// the Hanji: forcing the pair on is a compatibility projection of that,
    /// so every reader of the pair (auto-space, full-width punctuation, the
    /// nextword gates) behaves as today's hanji-first mode
    /// (`behavioral-invariants.md` §42). `.romanOnly` has no Hanji to lead with.
    /// CROSS-PLATFORM INVARIANT — mirrors ios `SettingsModels.swift`
    /// `CandidateDisplayMode.effectiveTranslateSwapped`, android
    /// `CandidateDisplayMode.kt`, windows `engine_settings.rs`. Drift causes
    /// silent divergence.
    func effectiveTranslateSwapped(stored: Bool) -> Bool {
        self == .combined || (stored && showsHanji)
    }

    /// Effective Annotate in Brackets for a stored flag — off only where there is no Hanji
    /// to bracket; `.combined` keeps it (`Hanji (romanization)`).
    func effectiveOutputBothScripts(stored: Bool) -> Bool {
        stored && showsHanji
    }

    /// Whether a typed punctuation key becomes full-width (`，` for `,`) for a
    /// stored swap flag — the stored flag masked like Annotate in Brackets, NOT the candidate
    /// projection above, which `.combined` forces on while the swap shortcut
    /// still picks the width. Mirrored on ios / android / windows beside
    /// `effectiveOutputBothScripts`.
    func effectiveFullWidthPunctuation(stored: Bool) -> Bool {
        stored && showsHanji
    }
}

/// Immutable snapshot of everything the engine needs to render a composition.
///
/// A snapshot rather than a set of getters because a single user intent can
/// issue several FFI calls (an `Append` is immediately followed by an
/// `EnterContinuous`), and those calls must agree: reading the settings twice
/// could straddle a change and render the two halves of one keystroke under
/// different rules.
struct EngineSettings: Equatable, Sendable {
    let inputMode: InputMode

    /// Word-boundary spacing inputs for the engine's `continuous_word_space`
    /// predicate (`docs/engine/continuous-input-ranking.md` §10.2). Both are
    /// `false` until PR5 ships the settings UI, but they are carried in the
    /// snapshot rather than hardcoded at the call sites because iOS's #380
    /// regression was exactly a call site that stopped passing the live pair.
    /// `isOutputBothScripts` is what separates hanji-first (no inter-segment
    /// space) from both-scripts (`hit (彼)` — space wanted); the swap flag is
    /// `true` for both, so one flag cannot express it.
    ///
    /// EFFECTIVE, not stored: under `candidateDisplayMode == .romanOnly` both
    /// read `false` whatever the user has stored, because a mode that shows
    /// and commits only romanization has no Hanji to lead with or to bracket.
    /// The stored values live on in `UserDefaults`
    /// (`SettingsStore.storedIsTranslateSwapped` / `storedIsOutputBothScripts`)
    /// and come back the moment the mode returns to `.sideBySide`. Under
    /// `.combined` the swap reads `true` whatever is stored — the Hanji cell
    /// comes first and is the `.primary` commit, the romanization cell beside
    /// it the `.alternate` one (`PresentedCandidate`) — while the bracket
    /// setting is read as stored (`SettingsStore.current` has the why). Every
    /// reader of "swap" — engine `AppConfig`, cell, document text, auto-space
    /// — reads THIS pair, never the stored one; full-width punctuation reads
    /// `isFullWidthPunctuation` instead.
    /// CROSS-PLATFORM INVARIANT — mirrors ios/Sources/TaigiKeyboard/Settings/EngineSettings.swift
    /// `isTranslateSwapped` / `isOutputBothScripts` (derived the same way) and
    /// the Windows `document.rs engine_settings()`. Drift changes what a
    /// romanization-only install commits.
    let isTranslateSwapped: Bool
    let isOutputBothScripts: Bool

    /// `CandidateDisplayMode.effectiveFullWidthPunctuation(stored:)` — read by
    /// `TaigiInputController.documentPunctuation` only.
    let isFullWidthPunctuation: Bool

    /// Whether the candidate window shows both scripts or the romanization
    /// alone. Sent to the engine as `AppConfig.candidate_display_mode`, which
    /// is what collapses same-romanization rows under `.romanOnly`
    /// (`engine/composing/src/dispatch.rs`, `engine/nextword/src/filter.rs`);
    /// on this side it selects the cell arm and derives the pair above.
    /// CROSS-PLATFORM INVARIANT — mirrors ios/Sources/TaigiKeyboard/Settings/EngineSettings.swift
    /// `candidateDisplayMode`, which defaults it to side-by-side. Drift changes
    /// what a fresh install's candidate cells show.
    let candidateDisplayMode: CandidateDisplayMode

    /// §34/S22 — when on, TL/POJ composing surfaces the preedit literal as the
    /// index-0 candidate so mixed-script writing commits the romanization in one keystroke, and
    /// Return on a fresh bar writes what was typed. The bridge inverts it into
    /// `FetchAtPos.literal_roman_candidate_disabled`.
    /// CROSS-PLATFORM INVARIANT — mirrors `isLiteralRomanCandidateEnabled` in
    /// ios/Sources/TaigiKeyboard/Settings/SharedSettings.swift and
    /// `literalRomanCandidateEnabled` in android/…/ime/core/PrefHelper.kt,
    /// both of which default it ON.
    /// Drift changes which candidate leads the list on a fresh install.
    let isLiteralRomanCandidateEnabled: Bool

    /// No Hyphens (`behavioral-invariants.md` §49) — sent as
    /// `AppConfig.hyphenless_roman` on the base config; no TPS layout here,
    /// so no fold.
    /// CROSS-PLATFORM INVARIANT — mirrors `isHyphenlessRomanEnabled` in
    /// ios/Sources/TaigiKeyboard/Settings/SharedSettings.swift and
    /// `hyphenlessRomanEnabled` in android/…/ime/core/PrefHelper.kt, both OFF.
    let isHyphenlessRomanEnabled: Bool

    /// ⁿ becomes ᴺ in capitals (`behavioral-invariants.md` §53) — the POJ nasal marker follows
    /// the case of the letters before it (`SIÂᴺ`); off, it is always `ⁿ`.
    /// Sent inverted as `AppConfig.force_lowercase_nasal_marker` on the base
    /// config.
    /// CROSS-PLATFORM INVARIANT — mirrors `isNasalMarkerUppercaseEnabled` in
    /// ios/Sources/TaigiKeyboard/Settings/SharedSettings.swift and
    /// `nasalMarkerUppercaseEnabled` in android/…/ime/core/PrefHelper.kt, all ON.
    let isNasalMarkerUppercaseEnabled: Bool

    /// Whether committing a candidate counts towards its ranking next time.
    /// Read on the write path only — the boost itself is always applied to
    /// whatever counts have been learned, so turning this off freezes the
    /// learned ranking rather than discarding it.
    /// CROSS-PLATFORM INVARIANT — mirrors ios/Sources/TaigiKeyboard/Settings/SharedSettings.swift:48,
    /// which defaults it ON. Drift changes whether a fresh install learns.
    let isFrequencyRecordingEnabled: Bool

    /// Whether the user's own dictionary contributes candidates. Gates the
    /// lookup itself, not just the display: with it off nothing is read from
    /// `custom_dictionary.db` (`FetchAtPos.custom_dictionary_disabled`).
    /// CROSS-PLATFORM INVARIANT — mirrors ios/Sources/TaigiKeyboard/Settings/SharedSettings.swift:51,
    /// which defaults it ON.
    let isCustomDictEnabled: Bool

    /// Which bundled dictionaries the engine may draw candidates from. Reaches
    /// the engine as `FetchAtPos.enabled_sources_bitmask` after the
    /// `compute_filters` op resolves it (`RustEngineBridge+Lexicon.swift`).
    let dictionarySources: DictionarySourceToggles

    /// What a fresh install types with. Every value matches the iOS and Android
    /// default for the same setting, so someone using two of the four platforms
    /// gets the same composition and the same candidate order out of the box.
    ///
    /// Hanji-first since 2026-09-18 (USER: "by default Hanji is output first, that is, Hanji is the
    /// title and romanization is the subtitle"): the stored swap is on, and the two effective
    /// fields are DERIVED from it under Hanji–Romanization Pairing the way `SettingsStore.current`
    /// derives them, so the snapshot cannot say one thing about the swap and
    /// another about the width.
    static let defaults: EngineSettings = {
        let storedSwap = true
        let mode = CandidateDisplayMode.sideBySide
        return EngineSettings(
            inputMode: .tl,
            isTranslateSwapped: mode.effectiveTranslateSwapped(stored: storedSwap),
            isOutputBothScripts: false,
            isFullWidthPunctuation: mode.effectiveFullWidthPunctuation(stored: storedSwap),
            candidateDisplayMode: mode,
            isLiteralRomanCandidateEnabled: true,
            isHyphenlessRomanEnabled: false,
            isNasalMarkerUppercaseEnabled: true,
            isFrequencyRecordingEnabled: true,
            isCustomDictEnabled: true,
            dictionarySources: .defaults,
        )
    }()
}

File: desktop/crates/taigi-desktop-core/src/settings/engine_settings.rs
//! The snapshot one engine operation reads. Port of
//! `macos/Sources/TaigiInputMethodCore/Settings/EngineSettings.swift` and
//! `DictionarySourceToggles.swift`; every default matches iOS / Android / macOS.

use super::choices::SettingChoice;

/// The romanization the user types. Desktop ships TL and POJ only — TPS is
/// out of scope on Windows as on macOS (`docs/architecture/windows-roadmap.md`
/// § Goal), which is why there is no `Tps` variant to fall through.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum InputMode {
    Tl,
    Poj,
}

impl InputMode {
    /// The `AppConfig.input_mode` wire spelling.
    pub fn wire(self) -> &'static str {
        self.raw()
    }

    /// The picker row's i18n key, as every other choice type carries one
    /// (`GeneralSettingsView.swift`).
    pub fn label_key(self) -> crate::strings::StringKey {
        use crate::strings::StringKey;
        match self {
            Self::Tl => StringKey::SettingsTlMode,
            Self::Poj => StringKey::SettingsPojMode,
        }
    }
}

impl SettingChoice for InputMode {
    const ALL: &'static [Self] = &[Self::Tl, Self::Poj];
    const DEFAULT: Self = Self::Tl;
    fn raw(self) -> &'static str {
        match self {
            Self::Tl => "tl",
            Self::Poj => "poj",
        }
    }
}

/// What a candidate cell shows: both scripts (the swap decides which leads),
/// the romanization alone, or both in ONE label led by the hanji (Hanji with Romanization,
/// `Combined`). Stored spellings and the per-mode rules are the same on every
/// platform (`SettingsModels.swift` `CandidateDisplayMode`).
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum CandidateDisplayMode {
    SideBySide,
    RomanOnly,
    Combined,
}

impl CandidateDisplayMode {
    /// Whether the cell shows any hanji — `false` only for `RomanOnly`.
    pub const fn shows_hanji(self) -> bool {
        !matches!(self, Self::RomanOnly)
    }

    /// Whether the swap shortcut writes the stored swap — exactly where hanji
    /// is on screen. Under `Combined` the cells are split per script, so the
    /// shortcut only picks the punctuation width (USER 2026-09-13: "Hanji with Romanization needs an
    /// isTranslateSwapped button"); `RomanOnly` leaves it inert and the stored
    /// swap waits for the way back.
    pub fn allows_swap_toggle(self) -> bool {
        self.shows_hanji()
    }

    /// Effective swap for a stored flag. `Combined` leads with — and commits —
    /// the hanji: forcing the pair on is a compatibility projection of that,
    /// so every reader of the pair (auto-space, the nextword gates) behaves
    /// as today's hanji-first mode (invariants §42); full-width punctuation
    /// reads `effective_full_width_punctuation` instead.
    /// `RomanOnly` has no hanji to lead with.
    /// CROSS-PLATFORM INVARIANT — mirrors macOS `EngineSettings.swift`
    /// `CandidateDisplayMode.effectiveTranslateSwapped`, iOS
    /// `SettingsModels.swift`, Android `CandidateDisplayMode.kt`.
    pub const fn effective_translate_swapped(self, stored: bool) -> bool {
        matches!(self, Self::Combined) || (stored && self.shows_hanji())
    }

    /// Effective Annotate in Brackets for a stored flag — off only where there is no hanji
    /// to bracket; `Combined` keeps it (`Hanji (romanization)`).
    pub fn effective_output_both_scripts(self, stored: bool) -> bool {
        stored && self.shows_hanji()
    }

    /// Whether a typed punctuation key becomes full-width (`，` for `,`) for a
    /// stored swap flag — the stored flag masked like Annotate in Brackets, NOT the
    /// candidate projection above, which `Combined` forces on while the swap
    /// shortcut still picks the width. Mirrored on macOS / iOS / Android
    /// beside `effective_output_both_scripts`.
    pub const fn effective_full_width_punctuation(self, stored: bool) -> bool {
        stored && self.shows_hanji()
    }

    /// The `AppConfig.candidate_display_mode` wire value. The engine reads
    /// it through `AppConfig::is_roman_only_display`, so only `RomanOnly`
    /// has to be exact; `SideBySide` is spelled out rather than left
    /// `Unspecified` so a build that sets the field is telling apart from
    /// one that never did. `Combined` has no engine reader either: a combined
    /// cell is distinct by its `(Hanji, romanization)` pair, so nothing collapses.
    pub fn wire(self) -> protos::engine::CandidateDisplayMode {
        match self {
            Self::SideBySide => protos::engine::CandidateDisplayMode::SideBySide,
            Self::RomanOnly => protos::engine::CandidateDisplayMode::RomanOnly,
            Self::Combined => protos::engine::CandidateDisplayMode::Combined,
        }
    }

    /// The picker row's i18n key.
    pub fn label_key(self) -> crate::strings::StringKey {
        use crate::strings::StringKey;
        match self {
            Self::SideBySide => StringKey::SettingsCandidateDisplayModeSideBySide,
            Self::RomanOnly => StringKey::SettingsCandidateDisplayModeRomanOnly,
            Self::Combined => StringKey::SettingsCandidateDisplayModeCombined,
        }
    }

    /// The mode after this one in picker order — what the
    /// `CycleCandidateDisplayMode` shortcut steps to: Pairing → Combined → Romanization Only →
    /// Pairing. CROSS-PLATFORM INVARIANT — mirrors macOS `CandidateDisplayMode.next`.
    pub fn next(self) -> Self {
        match self {
            Self::SideBySide => Self::Combined,
            Self::Combined => Self::RomanOnly,
            Self::RomanOnly => Self::SideBySide,
        }
    }
}

impl SettingChoice for CandidateDisplayMode {
    const ALL: &'static [Self] = &[Self::SideBySide, Self::Combined, Self::RomanOnly];
    /// Today's behaviour, byte for byte.
    const DEFAULT: Self = Self::SideBySide;
    fn raw(self) -> &'static str {
        match self {
            Self::SideBySide => "sideBySide",
            Self::RomanOnly => "romanOnly",
            Self::Combined => "combined",
        }
    }
}

/// Immutable snapshot of everything the engine needs to render a composition.
///
/// A snapshot rather than a set of getters because a single user intent can
/// issue several engine calls (an `Append` is immediately followed by an
/// `EnterContinuous`), and those calls must agree (`EngineSettings.swift:14-20`).
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct EngineSettings {
    pub input_mode: InputMode,
    /// Word-boundary spacing inputs for the engine's `continuous_word_space`
    /// predicate (`docs/engine/continuous-input-ranking.md` §10.2). Both are
    /// the EFFECTIVE values: the stored toggles AND-ed with `candidate_display_mode
    /// != RomanOnly`, and the swap forced true under `Combined`
    /// (`SettingsDocument::engine_settings`), never the raw document bools —
    /// the raw ones stay untouched so leaving either mode restores them.
    pub is_translate_swapped: bool,
    pub is_output_both_scripts: bool,
    /// `CandidateDisplayMode::effective_full_width_punctuation(stored)` —
    /// read by the TSF session's `document_punctuation` only.
    pub is_full_width_punctuation: bool,
    /// What a candidate cell shows; `AppConfig.candidate_display_mode`.
    /// CROSS-PLATFORM INVARIANT — mirrors
    /// `macos/Sources/TaigiInputMethodCore/Settings/EngineSettings.swift`
    /// `candidateDisplayMode`; every platform defaults to side-by-side.
    pub candidate_display_mode: CandidateDisplayMode,
    /// §34/S22 — inverted onto `FetchAtPos.literal_roman_candidate_disabled`.
    /// CROSS-PLATFORM INVARIANT — mirrors `isLiteralRomanCandidateEnabled`
    /// (`ios/.../SharedSettings.swift`) and `literalRomanCandidateEnabled`
    /// (`android/.../PrefHelper.kt`), both default ON.
    pub is_literal_roman_candidate_enabled: bool,
    /// No Hyphens (`behavioral-invariants.md` §49) — `AppConfig.hyphenless_roman`
    /// on the base config; no TPS layout here, so no fold.
    /// CROSS-PLATFORM INVARIANT — mirrors `isHyphenlessRomanEnabled`
    /// (`macos/.../EngineSettings.swift`, `ios/.../SharedSettings.swift`) and
    /// `hyphenlessRomanEnabled` (`android/.../PrefHelper.kt`), all OFF.
    pub is_hyphenless_roman_enabled: bool,
    /// ⁿ becomes ᴺ in capitals (`behavioral-invariants.md` §53) — the POJ nasal marker follows
    /// the case of the letters before it (`SIÂᴺ`); off, always `ⁿ`. Sent
    /// inverted as `AppConfig.force_lowercase_nasal_marker` on the base config.
    /// CROSS-PLATFORM INVARIANT — mirrors `isNasalMarkerUppercaseEnabled`
    /// (`macos/.../EngineSettings.swift`, `ios/.../SharedSettings.swift`) and
    /// `nasalMarkerUppercaseEnabled` (`android/.../PrefHelper.kt`), all ON.
    pub is_nasal_marker_uppercase_enabled: bool,
    /// Read on the write path only; the boost always applies to whatever was
    /// learned. CROSS-PLATFORM INVARIANT — `SharedSettings.swift:48` (ON).
    pub is_frequency_recording_enabled: bool,
    /// Gates the custom-dictionary lookup itself: off means the engine reads
    /// no custom rows (`FetchAtPos.custom_dictionary_disabled`). CROSS-PLATFORM INVARIANT —
    /// `SharedSettings.swift:51` (ON).
    pub is_custom_dict_enabled: bool,
    pub dictionary_sources: DictionarySourceToggles,
}

impl EngineSettings {
    /// What a fresh install types with. Every value matches the iOS, Android
    /// and macOS default for the same setting (`EngineSettings.swift:79-88`).
    /// A `const` so the settings keys (`keys.rs`) can read their defaults
    /// from it rather than restate them.
    ///
    /// Hanji-first (USER 2026-09-18): the stored swap is on, and the two
    /// effective fields are DERIVED from it under side-by-side the way
    /// `SettingsDocument::engine_settings` derives them, so the snapshot
    /// cannot say one thing about the swap and another about the width.
    pub const DEFAULT: Self = {
        const STORED_SWAP: bool = true;
        const MODE: CandidateDisplayMode = CandidateDisplayMode::SideBySide;
        Self {
            input_mode: InputMode::Tl,
            is_translate_swapped: MODE.effective_translate_swapped(STORED_SWAP),
            is_output_both_scripts: false,
            is_full_width_punctuation: MODE.effective_full_width_punctuation(STORED_SWAP),
            candidate_display_mode: MODE,
            is_literal_roman_candidate_enabled: true,
            is_hyphenless_roman_enabled: false,
            is_nasal_marker_uppercase_enabled: true,
            is_frequency_recording_enabled: true,
            is_custom_dict_enabled: true,
            dictionary_sources: DictionarySourceToggles::DEFAULT,
        }
    };
}

impl Default for EngineSettings {
    fn default() -> Self {
        Self::DEFAULT
    }
}

/// Which bundled dictionaries the user has switched on, in the shape the
/// engine's `compute_filters` op reads them. Field order mirrors
/// `engine/protos/proto/lexicon.proto::DictionaryToggles`.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct DictionarySourceToggles {
    /// 教育部臺灣台語常用詞辭典 (settings key `moeDictEnabled`).
    pub kautian: bool,
    /// 台語新詞辭庫.
    pub taigitv: bool,
    /// iTaigi 華台對照典.
    pub itaigi: bool,
    /// 台灣植物名彙.
    pub sitbut: bool,
    /// 台華線頂對照典.
    pub taihoa: bool,
    /// 台日大辭典.
    pub taijit: bool,
    /// 台語工藝詞庫.
    pub kungge: bool,
    /// 學科術語辭典.
    pub stti: bool,
    /// 腔口補充資料.
    pub khpoo: bool,
    /// Variant Characters (異用字).
    pub variant: bool,
    /// Conventional Characters (在來字).
    pub khiin: bool,
    /// LKK Han-Lo Recommended Characters.
    pub lkk: bool,
    /// Supplementary Word List (詞庫增補檔案).
    pub dev: bool,
    /// Always populated: an absent subcollection message tells the engine to
    /// skip the gate (`lexicon.proto` `DictionaryToggles.kautian_subcoll`).
    pub kautian_subcollections: KautianSubcollections,
}

impl DictionarySourceToggles {
    /// CROSS-PLATFORM INVARIANT — mirrors `ios/.../SharedSettings.swift:53-66`
    /// and `macos/.../DictionarySourceToggles.swift:99-114`.
    pub const DEFAULT: Self = Self {
        kautian: true,
        taigitv: true,
        itaigi: false,
        sitbut: false,
        taihoa: false,
        taijit: false,
        kungge: true,
        stti: true,
        khpoo: true,
        variant: false,
        khiin: false,
        lkk: true,
        dev: true,
        kautian_subcollections: KautianSubcollections::DEFAULT,
    };
}

impl Default for DictionarySourceToggles {
    fn default() -> Self {
        Self::DEFAULT
    }
}

/// The per-subcollection state of the kautian source. Order mirrors
/// `KautianSubcollToggles` (`lexicon.proto`), the `config.yaml`
/// `dialect_columns` order.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct KautianSubcollections {
    pub accent_lukang: bool,
    pub accent_sansia: bool,
    pub accent_taipak: bool,
    pub accent_gilan: bool,
    pub accent_tainan: bool,
    pub accent_kaohsiung: bool,
    pub accent_kinmen: bool,
    pub accent_makung: bool,
    pub accent_sintik: bool,
    pub accent_taichung: bool,
    pub name_appendix: bool,
}

impl KautianSubcollections {
    /// CROSS-PLATFORM INVARIANT — every subcollection defaults ON
    /// (`ios/.../SharedSettings.swift:74-84`).
    pub const DEFAULT: Self = Self {
        accent_lukang: true,
        accent_sansia: true,
        accent_taipak: true,
        accent_gilan: true,
        accent_tainan: true,
        accent_kaohsiung: true,
        accent_kinmen: true,
        accent_makung: true,
        accent_sintik: true,
        accent_taichung: true,
        name_appendix: true,
    };
}

impl Default for KautianSubcollections {
    fn default() -> Self {
        Self::DEFAULT
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn display_mode_cycle_follows_the_picker_and_returns_in_three_steps() {
        // trace: `SettingChoice::ALL` = [SideBySide, Combined, RomanOnly];
        // `next` walks that ring, so the third step is back at the start.
        let start = CandidateDisplayMode::SideBySide;
        let one = start.next();
        let two = one.next();
        assert_eq!(one, CandidateDisplayMode::Combined);
        assert_eq!(two, CandidateDisplayMode::RomanOnly);
        assert_eq!(two.next(), start);
        for (index, mode) in CandidateDisplayMode::ALL.iter().enumerate() {
            let successor =
                CandidateDisplayMode::ALL[(index + 1) % CandidateDisplayMode::ALL.len()];
            assert_eq!(
                mode.next(),
                successor,
                "{mode:?} steps off the picker order"
            );
        }
    }
}
